### PR TITLE
Various improvements and additions to Python sample code

### PR DESCRIPTION
**Pull request by einarnn**
In summary:

- Some refactoring of existing code to tidy certain things up
- More logging and verbose options
- Updated print statements to **mostly** print just machine-readable data to facilitate piping data onwards (e.g. pass session query results on to a utility like `jq`)
- Make it easier to use username/password and accept self-signed certs from ISE
    - Warning generated that this is not desirable for production deployments
- General purpose subscription sample, `px_subscribe.py`, that has a number of options, including subscribing to all pxGrid nodes in a deployment and deduplicating **session events only**
- Samples of getting TrustSec objects and subscribing to TrustSec event feeds (topic availability ISE version-dependent)

### CODE REVIEW

**Review by alei121 — COMMENTED** [review]
So far, I have been trying to keep the samples simple and consistent across different languages.
But I do see a need to add more specific examples.
We should probably reorganize the directory structure to accommodate this later.

File: python/pxgrid.py [hunk 1]
```diff
@@ -1,48 +1,58 @@
...
+import logging
+
+logger = logging.getLogger(__name__)
 
 
 class PxgridControl:
     def __init__(self, config):
         self.config = config
+        if self.config.verbose:
+            handler = logging.StreamHandler()
```
[the remark this answers is not in the document]
Comment (einarnn) [reply python/pxgrid.py:13]: The main functions in various scripts only set for the main module itself and for the `stomp` and `ws_stomp` modules, not for the `pxgrid` module. I will centralize to the main modules for now and take out of `pxgrid.py`

[the remark this answers is not in the document]
Comment (einarnn) [reply python/pxgrid.py]: Ok, updates for this committed.

File: python/config.py [hunk 2]
```diff
@@ -19,28 +21,106 @@ def __init__(self):
...
+            if self.config.clientcert is not None:
+                self.__ssl_context.load_cert_chain(
+                    certfile=self.config.clientcert,
+                    keyfile=self.config.clientkey,
+                    password=self.config.clientkeypassword)
+            if self.config.servercert:
+                self.__ssl_context.load_verify_locations(cafile=self.config.servercert)
+            else:
+                warnings.warn("check_hostname and cert not used; unsafe for production use")
+                self.__ssl_context.check_hostname = False
```
[the remark this answers is not in the document]
Comment (einarnn) [reply python/config.py:131]: Do you mean a specific option to turn off the warning or an option to disallow `self.__ssl_context.check_hostname = False`?

**Comment by einarnn** [comment]
> So far, I have been trying to keep the samples simple and consistent across different languages.
> But I do see a need to add more specific examples.
> We should probably reorganize the directory structure to accommodate this later.

While I think it would be great to keep all the samples with common functionality, I'm not sure how practical it is. For example, I'd find it challenging to keep the Java & Go sample apps up to date with the Python in terms of the time commitment.

How would reorganizing the directory structure help?

**Comment by ronakg** [comment]
> > So far, I have been trying to keep the samples simple and consistent across different languages.
> > But I do see a need to add more specific examples.
> > We should probably reorganize the directory structure to accommodate this later.
> 
> While I think it would be great to keep all the samples with common functionality, I'm not sure how practical it is. For example, I'd find it challenging to keep the Java & Go sample apps up to date with the Python in terms of the time commitment.
> 
> How would reorganizing the directory structure help?

Yes, we should reorganize the directory structure to make sure that there's consistency. We should have basic examples which are available in all 3 languages. The changes introduced in this PR can be called advanced examples and these may or may not be available in all 3 languages.

File: python/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alei121) [reply python/config.py:131]: For the server certificate option, if not specified, we should rely on the system's CA. I think that's the default behavior of the ssl_context.
And then add an new option to disable the certificate check. Like "wget --no-check-certificate" or "curl --insecure". If this option is not specified, the default should verify certificates and hostname.

**Comment by einarnn** [comment]
> > > So far, I have been trying to keep the samples simple and consistent across different languages.
> > > But I do see a need to add more specific examples.
> > > We should probably reorganize the directory structure to accommodate this later.
> > 
> > 
> > While I think it would be great to keep all the samples with common functionality, I'm not sure how practical it is. For example, I'd find it challenging to keep the Java & Go sample apps up to date with the Python in terms of the time commitment.
> > How would reorganizing the directory structure help?
> 
> Yes, we should reorganize the directory structure to make sure that there's consistency. We should have basic examples which are available in all 3 languages. The changes introduced in this PR can be called advanced examples and these may or may not be available in all 3 languages.

I'm fine with that in theory. Worth noting that I've also tweaked what you call the basic examples. I haven't changed the functionality, but the common classes/functions have been modified. This common code needs to be used by both "basic" and "advanced" examples, and so there probably needs to be a larger overall refactoring.

Do you feel this refactoring needs to happen before this code can be committed?

File: python/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (einarnn) [reply python/config.py:131]: Ahh, so you're asking for the equivalent of curl's `-k` flag. Sure. I can add that.

[the remark this answers is not in the document]
Comment (einarnn) [reply python/config.py:131]: I'll add these flags:

```
 -k, --insecure      Allow insecure server connections when using SSL
```

Good?

**Comment by einarnn** [comment]
@alei121 @ronakg -- any chance we can get this PR merged? As far as I can tell I've addressed all comments made.

Cheers,

Einar

**Comment by einarnn** [comment]
These samples will be added to a standalone repo focused on pxGrid 2.0 Python samples.

**Pull request closed without merging**